pkg/startup/autodetection/ipv4/poolselector.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2020 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package ipv4
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+)
+
+const (
+	FALLBACK_IPPOOL_TEMPLATE = "172.%d.0.0/16"
+	FALLBACK_IPPOOL_MIN      = 16
+	FALLBACK_IPPOOL_MAX      = 31
+)
+
+type poolSelector struct {
+	actualPool    *net.IPNet
+	poolCounter   int
+	hostAddresses []net.IPNet
+}
+
+var hostIPAddressRetriever func() ([]net.IPNet, error) = retrieveIPAddresses
+
+// GetDefaultIPv4Pool detects host interfaces and selects default IP pool without overlapping
+func GetDefaultIPv4Pool(preferedPool string) string {
+	log.Debug("Auto-detecting IPv4 interfaces to select default pool")
+
+	_, preferedNet, err := net.ParseCIDR(preferedPool)
+	if err != nil {
+		log.Errorf("Unable to parse prefered pool: '%s'", err.Error())
+		return preferedPool
+	}
+
+	hostAddresses, err := hostIPAddressRetriever()
+	if err != nil {
+		log.Errorf("Unable to retrieve host network interfaces: '%s'", err.Error())
+		return preferedPool
+	}
+
+	selector := poolSelector{
+		actualPool:    preferedNet,
+		hostAddresses: hostAddresses,
+	}
+
+	selectedPool, err := selector.selectDefaultPool()
+	if err != nil {
+		log.Errorf("Unable to select default pool: '%s'", err.Error())
+		return preferedPool
+	}
+
+	return selectedPool
+}
+
+func retrieveIPAddresses() ([]net.IPNet, error) {
+	linkAddresses, err := netlink.AddrList(nil, netlink.FAMILY_V4)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to list addresses")
+	}
+
+	resp := make([]net.IPNet, len(linkAddresses))
+	for _, addr := range linkAddresses {
+		_, netAddr, err := net.ParseCIDR(addr.IPNet.String())
+		if err != nil {
+			return nil, errors.Wrapf(err, "Unable to parse address '%s'", addr.String())
+		}
+
+		resp = append(resp, *netAddr)
+	}
+
+	return resp, nil
+}
+
+func (s *poolSelector) selectDefaultPool() (string, error) {
+	for _, addr := range s.hostAddresses {
+		if s.actualPool.Contains(addr.IP) {
+			if err := s.hopToNextPool(); err != nil {
+				return "", errors.Wrap(err, "Unable to find unused IP pool")
+			}
+			return s.selectDefaultPool()
```

Comment:
I think the recursion here makes this code a bit harder to understand. How do you feel about unpicking this so that it doesn't use recursion?

Here's some pseudocode:

```go
func findAvailableCIDR(hostAddrs []net.IPNet) net.IPNet {
  // Check if the preferred pool overlaps with any host addresses. If it doesn't then just use that.
  if overlaps := doesPoolOverlap(preferredPool, hostAddrs); !overlaps {
    return preferredPool
  }

  // The preferred pool overlapped. Look for another one.
  for i = 16 -> i = 32 {
    pooltoTry = fmt.Sprintf(TEMPLATE, i)
    if overlaps = doesPoolOverlap(pooltoTrye, hostAddrs); !overlaps {
      // Found one, use it.
      return poolToTry
    }
  }

  // We couldn't find a pool. For compatibility reasons, return the preferred pool to match existing behavior.
  return preferredPool
}
```

Reply:
I can agree. It is more procedural but anyway Go is a procedural language :D